trunk/lmbench3/doc/lat_mem_rd.8
.\" $Id$
.TH LAT_MEM_RD 8 "$Date$" "(c)1994 Larry McVoy" "LMBENCH"
.SH NAME
lat_mem_rd \- memory read latency benchmark
.SH SYNOPSIS
.B lat_mem_rd 
.I size_in_megabytes
.I stride [stride stride...]
.SH DESCRIPTION
.B lat_mem_rd
measures memory read latency for varying memory sizes and strides.  The
results are reported in nanoseconds per load and have been verified
accurate to within a few nanoseconds on an SGI Indy.
.LP
The
entire memory hierarchy is measured, including onboard cache latency
and size, external cache latency and size, main memory latency, and TLB
miss latency.
.LP
Only data accesses are measured; the instruction cache is not measured.
.LP
The benchmark runs as two nested loops.  The outer loop is the stride size.
The inner loop is the array size.  For each array size, the benchmark
creates a ring of pointers that point forward one stride.  Traversing the
array is done by
.sp
.ft CB
	p = (char **)*p;
.ft
.sp
in a for loop (the over head of the for loop is not significant; the loop is 
an unrolled loop 1000 loads long).  The loop stops after doing a million loads.
.LP
The size of the array varies from 512 bytes to (typically) eight megabytes.
For the small sizes, the cache will have an effect, and the loads will be
much faster.  This becomes much more apparent when the data is plotted.
.SH OUTPUT
Output format is intended as input to \fBxgraph\fP or some similar program
(I use a perl script that produces pic input).
There is a set of data produced for each stride.  The data set title
is the stride size and the data points are the array size in megabytes 
(floating point value) and the load latency over all points in that array.
.SH "INTERPRETING THE OUTPUT"
The output is best examined in a graph where you typically get a graph
that has four plateaus.  The graph should plotted in log base 2 of the
array size on the X axis and the latency on the Y axis.  Each stride
is then plotted as a curve.  The plateaus that appear correspond to 
the onboard cache (if present), external cache (if present), main
memory latency, and TLB miss latency.
.LP
As a rough guide, you may be able to extract the latencies of the
various parts as follows, but you should really look at the graphs,
since these rules of thumb do not always work (some systems do not
have onboard cache, for example).
.IP "onboard cache" 16
Try stride of 128 and array size of .00098.
.IP "external cache" 
Try stride of 128 and array size of .125.
.IP "main memory"
Try stride of 128 and array size of 8.
.IP "TLB miss"
Try the largest stride and the largest array.
.SH BUGS
This program is dependent on the correct operation of
.BR mhz (8).
If you are getting numbers that seem off, check that 
.BR mhz (8)
is giving you a clock rate that you believe.
.SH ACKNOWLEDGEMENT
Funding for the development of
this tool was provided by Sun Microsystems Computer Corporation.
.SH "SEE ALSO"
lmbench(8).


lmbench2/src/lat_mem_rd.c
/*
 * lat_mem_rd.c - measure memory load latency
 *
 * usage: lat_mem_rd size-in-MB stride [stride ...]
 *
 * Copyright (c) 1994 Larry McVoy.  Distributed under the FSF GPL with
 * additional restriction that results may published only if
 * (1) the benchmark is unmodified, and
 * (2) the version in the sccsid below is included in the report.
 * Support for this development by Sun Microsystems is gratefully acknowledged.
 */
char	*id = "$Id$\n";

#include "bench.h"
#define N       1000000	/* Don't change this */
#define STRIDE  (512/sizeof(char *))
#define	MEMTRIES	4
#define	LOWER	512
void	loads(char *addr, size_t range, size_t stride);
int	step(int k);

int
main(int ac, char **av)
{
	size_t	len;
	size_t	range;
	size_t	stride;
	int	i;
        char   *addr;

        len = atoi(av[1]) * 1024 * 1024;
        addr = (char *)malloc(len);

	if (av[2] == 0) {
		fprintf(stderr, "\"stride=%d\n", STRIDE);
		for (range = LOWER; range <= len; range = step(range)) {
			loads(addr, range, STRIDE);
		}
	} else {
		for (i = 2; i < ac; ++i) {
			stride = bytes(av[i]);
			fprintf(stderr, "\"stride=%d\n", stride);
			for (range = LOWER; range <= len; range = step(range)) {
				loads(addr, range, stride);
			}
			fprintf(stderr, "\n");
		}
	}
	return(0);
}

void
loads(char *addr, size_t range, size_t stride)
{
	register char **p = 0 /* lint */;
	size_t	i;
	int	tries = 0;
	int	result = 0x7fffffff;
	double	time;

     	if (stride & (sizeof(char *) - 1)) {
		printf("lat_mem_rd: stride must be aligned.\n");
		return;
	}
	
	if (range < stride) {
		return;
	}

	/*
	 * First create a list of pointers.
	 *
	 * This used to go forwards, we want to go backwards to try and defeat
	 * HP's fetch ahead.
	 *
	 * We really need to do a random pattern once we are doing one hit per 
	 * page.
	 */
	for (i = range - stride; i >= 0; i -= stride) {
		char	*next;

		p = (char **)&addr[i];
		if (i < stride) {
			next = &addr[range - stride];
		} else {
			next = &addr[i - stride];
		}
		*p = next;
	}

	/*
	 * Now walk them and time it.
	 */
        for (tries = 0; tries < MEMTRIES; ++tries) {
                /* time loop with loads */
#define	ONE	p = (char **)*p;
#define	FIVE	ONE ONE ONE ONE ONE
#define	TEN	FIVE FIVE
#define	FIFTY	TEN TEN TEN TEN TEN
#define	HUNDRED	FIFTY FIFTY
		i = N;
                start(0);
                while (i >= 1000) {
			HUNDRED
			HUNDRED
			HUNDRED
			HUNDRED
			HUNDRED
			HUNDRED
			HUNDRED
			HUNDRED
			HUNDRED
			HUNDRED
			i -= 1000;
                }
		i = stop(0,0);
		use_pointer((void *)p);
		if (i < result) {
			result = i;
		}
	}
	/*
	 * We want to get to nanoseconds / load.  We don't want to
	 * lose any precision in the process.  What we have is the
	 * milliseconds it took to do N loads, where N is 1 million,
	 * and we expect that each load took between 10 and 2000
	 * nanoseconds.
	 *
	 * We want just the memory latency time, not including the
	 * time to execute the load instruction.  We allow one clock
	 * for the instruction itself.  So we need to subtract off
	 * N * clk nanoseconds.
	 *
	 * lmbench 2.0 - do the subtration later, in the summary.
	 * Doing it here was problematic.
	 *
	 * XXX - we do not account for loop overhead here.
	 */
	time = (double)result;
	time *= 1000.;				/* convert to nanoseconds */
	time /= (double)N;			/* nanosecs per load */
	fprintf(stderr, "%.5f %.3f\n", range / (1024. * 1024), time);
}

int
step(int k)
{
	if (k < 1024) {
		k = k * 2;
        } else if (k < 4*1024) {
		k += 1024;
        } else if (k < 32*1024) {
		k += 2048;
        } else if (k < 64*1024) {
		k += 4096;
        } else if (k < 128*1024) {
		k += 8192;
        } else if (k < 256*1024) {
		k += 16384;
        } else if (k < 512*1024) {
		k += 32*1024;
	} else if (k < 4<<20) {
		k += 512 * 1024;
	} else if (k < 8<<20) {
		k += 1<<20;
	} else if (k < 20<<20) {
		k += 2<<20;
	} else {
		k += 10<<20;
	}
	return (k);
}
